x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/TextExpansionQueryBuilder.java

```diff
@@ -0,0 +1,266 @@
+import org.elasticsearch.xpack.core.ml.inference.results.WarningInferenceResults;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.SlimConfigUpdate;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.core.ClientHelper.ML_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+
+public class TextExpansionQueryBuilder extends AbstractQueryBuilder<TextExpansionQueryBuilder> {
+
+    public static final String NAME = "text_expansion";
+    public static final ParseField MODEL_TEXT = new ParseField("model_text");
+    public static final ParseField MODEL_ID = new ParseField("model_id");
+
+    private static final String DEFAULT_MODEL_ID = "slim";
+
+    private final String fieldName;
+    private final String modelText;
+    private final String modelId;
+    private SetOnce<SlimResults> weightedTokensSupplier;
+    private QueryBuilder termsQuery;
+
+    public TextExpansionQueryBuilder(String fieldName, String modelText, String modelId) {
+        if (fieldName == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a fieldName");
+        }
+        if (modelText == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a " + MODEL_TEXT.getPreferredName() + " value");
+        }
+
+        this.fieldName = fieldName;
+        this.modelText = modelText;
+        this.modelId = modelId == null ? DEFAULT_MODEL_ID : modelId;
+    }
+
+    public TextExpansionQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.modelText = in.readString();
+        this.modelId = in.readString();
+        this.termsQuery = in.readOptionalNamedWriteable(QueryBuilder.class);
+    }
+
+    private TextExpansionQueryBuilder(TextExpansionQueryBuilder other, SetOnce<SlimResults> weightedTokensSupplier) {
+        this.fieldName = other.fieldName;
+        this.modelText = other.modelText;
+        this.modelId = other.modelId;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.weightedTokensSupplier = weightedTokensSupplier;
+    }
+
+    private TextExpansionQueryBuilder(TextExpansionQueryBuilder other, QueryBuilder rewrittenTermsQuery) {
+        this.fieldName = other.fieldName;
+        this.modelText = other.modelText;
+        this.modelId = other.modelId;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.termsQuery = rewrittenTermsQuery;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersion.V_8_8_0;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
```

Comment:
You should throw here if `weightedTokensSupplier != null`. Attempting to serialize in that state is bad. It shouldn't happen, but its better to be safe.